### PR TITLE
Limit number of 2D filtering outputs stored in memory

### LINKED ISSUE #73: Investigate memory usage
There have been a few reports of cellfinder requiring far too much RAM (100GB+). This seems to be new and should be investigated.

### PR DESCRIPTION
**Pull request by dstansby**
After running into a bit of a dead end on `dask` (see https://github.com/brainglobe/cellfinder-core/issues/118#issuecomment-1497513776 for more info), I had a re-think and brainwave on how to solve https://github.com/brainglobe/cellfinder-core/issues/73 and might have come up with a plan...

The idea is to create a lock for each plane in the input data, and only release the lock on 2D filtering for the next `n_procs` planes when the previous plane has started being processed by the 3D filter. This replicates the behaviour before version 0.3.0, and fixes https://github.com/brainglobe/cellfinder-core/issues/73

### CODE REVIEW

**Comment by adamltyson** [comment]
Seems like a good idea, but #145 is preventing me from testing it properly.

**Review by alessandrofelder — CHANGES_REQUESTED** [review]
I found a small edge case that needs guarding against, otherwise looks good. Some wondering about tests and locks to consider too :thinking: 

I ran into issues with real-life data and was about to open #149 and #158 when I realised someone had already :facepalm: 

I found an issue #159 with numba further down the line from #158 that we can tackle separately.

If I pass a slice of real-life data that does not require splitting the result looks sensible to me.

File: src/cellfinder_core/detect/detect.py [hunk 1]
```diff
@@ -144,31 +148,65 @@ def main(
...
+
+def _run_func_with_lock(func, arg, lock: Lock):
+    """
+    Run a function after acquiring a lock.
+    """
+    lock.acquire(blocking=True)
+    return func(arg)
+
+
+def _map_with_locks(
```
[the remark this answers is not in the document]
Comment (dstansby) [reply src/cellfinder_core/detect/detect.py:187]: I've added a basic test, but couldn't come up with anything much more than your suggestion

File: src/cellfinder_core/detect/detect.py [hunk 2]
```diff
@@ -144,31 +148,65 @@ def main(
...
-                mp_tile_processor.get_tile_mask, args=(plane,)
-            )
-            async_results.put(res)
+    with mp_ctx.Pool(n_ball_procs) as worker_pool:
+        async_results, locks = _map_with_locks(
+            mp_tile_processor.get_tile_mask, signal_array, worker_pool
+        )
+
+        # Release the first set of locks for the 2D filtering
+        for i in range(n_ball_procs + ball_z_size):
```
[the remark this answers is not in the document]
Comment (dstansby) [reply src/cellfinder_core/detect/detect.py:159]: It does indeed fix it, and I've added a test 👍

File: src/cellfinder_core/detect/detect.py [hunk 3]
```diff
@@ -16,13 +16,15 @@
 import multiprocessing
 from datetime import datetime
 from queue import Queue
-from typing import Callable, List, Optional, Tuple, Union
+from threading import Lock
```
[the remark this answers is not in the document]
Comment (dstansby) [reply src/cellfinder_core/detect/detect.py:19]: [`multiprocessing.Manager.Lock()`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.managers.SyncManager.Lock) returns `threading.Lock()` instance, so I think this has to stay as is.

**Review by alessandrofelder — APPROVED** [review]
Cool!
